pkg/apis/admissionregistration/validation/validation.go

```diff
@@ -1002,28 +1058,28 @@ func validateMatchConditionsExpression(expression string, opts validationOptions
 	if opts.preexistingExpressions.matchConditionExpressions.Has(expression) {
 		envType = environment.StoredExpressions
 	}
-	return validateCELCondition(&matchconditions.MatchCondition{
+	return validateCELCondition(statelessCELCompiler, &matchconditions.MatchCondition{
 		Expression: expression,
 	}, plugincel.OptionalVariableDeclarations{
 		HasParams:     opts.allowParamsInMatchConditions,
 		HasAuthorizer: true,
 	}, envType, fldPath)
 }
 
-func validateMessageExpression(expression string, opts validationOptions, fldPath *field.Path) field.ErrorList {
+func validateMessageExpression(compiler plugincel.Compiler, expression string, opts validationOptions, fldPath *field.Path) field.ErrorList {
 	envType := environment.NewExpressions
 	if opts.preexistingExpressions.validationMessageExpressions.Has(expression) {
 		envType = environment.StoredExpressions
 	}
-	return validateCELCondition(&validatingadmissionpolicy.MessageExpressionCondition{
+	return validateCELCondition(compiler, &validatingadmissionpolicy.MessageExpressionCondition{
 		MessageExpression: expression,
 	}, plugincel.OptionalVariableDeclarations{
 		HasParams:     opts.allowParamsInMatchConditions,
 		HasAuthorizer: false,
 	}, envType, fldPath)
 }
 
-func validateAuditAnnotation(meta metav1.ObjectMeta, v *admissionregistration.AuditAnnotation, paramKind *admissionregistration.ParamKind, opts validationOptions, fldPath *field.Path) field.ErrorList {
+func validateAuditAnnotation(compiler plugincel.Compiler, meta metav1.ObjectMeta, v *admissionregistration.AuditAnnotation, paramKind *admissionregistration.ParamKind, opts validationOptions, fldPath *field.Path) field.ErrorList {
```

Comment:
is compiler used in this function?

Reply:
Yes it is! `AuditAnnotation`.`ValueExpression` indeed supports variable composition.